src/fsharp/FSharp.Core/printf.fs

```diff
@@ -157,598 +224,238 @@ module internal PrintfImpl =
-                        env
-                    next env : 'Tail
-                )
-            )
+                | StepStar2(prefix, conv) -> 
+                    env.WriteSkipEmpty prefix
+                    let star1 = args.[argIndex] :?> int
+                    argIndex <- argIndex + 1
+                    let star2 = args.[argIndex] :?> int
+                    argIndex <- argIndex + 1
+                    let arg1 = args.[argIndex]
+                    argIndex <- argIndex + 1
+                    env.Write (conv arg1 star1 star2)
+
+                | StepPercentStar2(prefix) -> 
+                    env.WriteSkipEmpty prefix
+                    //let _star1 = args.[argIndex] :?> int
+                    argIndex <- argIndex + 2
+                    env.Write("%")
+    
+            env.Finish()
+
+    /// Type of results produced by specialization.
+    ///
+    /// This is a function that accepts a thunk to create PrintfEnv on demand (at the very last
+    /// appliction of an argument) and returns a concrete instance of an appriate curried printer.
+    ///
+    /// After all arguments are collected, specialization obtains concrete PrintfEnv from the thunk
+    /// and uses it to output collected data.
+    type PrintfFuncContext<'State, 'Residue, 'Result> = unit -> (obj list * PrintfEnv<'State, 'Residue, 'Result>)
+    type PrintfFuncFactory<'Printer, 'State, 'Residue, 'Result> = PrintfFuncContext<'State, 'Residue, 'Result> -> 'Printer
 
-        static member ChainedFastStart4<'A, 'B, 'C, 'D, 'Tail>
-            (
-                conv1, s1, conv2, s2, conv3, s3, conv4,
-                next
-            ) =
-            (fun (env: unit -> PrintfEnv<'State, 'Residue, 'Result>) ->
-                (fun (a: 'A) (b: 'B) (c: 'C) (d: 'D)->
-                    let env() = 
-                        let env = env()
-                        Utils.Write(env, (conv1 a), s1, (conv2 b), s2, (conv3 c), s3, (conv4 d))
-                        env
-                    next env : 'Tail
-                )
-            )
+    [<Literal>]
+    let MaxArgumentsInSpecialization = 3
 
-        static member Chained5<'A, 'B, 'C, 'D, 'E, 'Tail>
-            (
-                s0, conv1, s1, conv2, s2, conv3, s3, conv4, s4, conv5,
-                next
-            ) =
-            (fun (env: unit -> PrintfEnv<'State, 'Residue, 'Result>) ->
-                (fun (a: 'A) (b: 'B) (c: 'C) (d: 'D) (e: 'E)->
-                    let env() = 
-                        let env = env()
-                        Utils.Write(env, s0, (conv1 a), s1, (conv2 b), s2, (conv3 c), s3, (conv4 d), s4, (conv5 e))
-                        env
-                    next env : 'Tail
-                )
+    type Specializations<'State, 'Residue, 'Result>() =
+     
+        static let finalizeSteps steps = steps |> List.rev |> List.toArray
+        static let finalizeArgs args = args |> List.rev |> List.toArray
```

Comment:
Would it not be possible to construct the array as an in-order array being passed in rather than do this?

Reply:
It isn't possible to do the args this way, I've noted why in the comment